### PR TITLE
fix: fix broken canonicity in IFE processing

### PR DESCRIPTION
**Pull request by pgebal**
Closes https://github.com/omisego/security-issues/issues/13

### CODE REVIEW

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 1]
```diff
@@ -80,18 +75,25 @@ library PaymentProcessInFlightExit {
...
+                        withdrawFromVault(self, withdrawal);
+                        emit InFlightExitOutputWithdrawn(exitId, i);
+                    }
                 }
-            }
 
-            flagOutputsWhenCanonical(self.framework, exit, token);
+                flagOutputsWhenCanonical(self.framework, exit, token, exitId);
+            }
+            //TODO: possibly emit an event that output exit was skipped
```
Comment (pgebal) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: I don't think if it's needed, buy other opinions are welcomed.

File: plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py [hunk 2]
```diff
@@ -817,3 +817,40 @@ def start_ife_piggyback_and_process(spend_id, owner, testlang):
...
+    testlang.process_exits(token.address, 0, 1)
+    testlang.process_exits(NULL_ADDRESS, 0, 1)
+
+    # caroline exits with the tokens (previously owned by alice)
+    caroline_token_balance = token.balanceOf(caroline.address)
+    assert caroline_token_balance == caroline_token_balance_before + amount_token
+
+    # but she can not exit with Eth
+    caroline_eth_balance = testlang.get_balance(caroline)
+    # TODO: piggyback bond is still returned, I am not sure if that's good or bad
```
Comment (pgebal) [inline plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py]: We still return piggyback bond in that case. That's not harmful. Someone have other opinions?

**Comment by pgebal** [comment]
Looks like I made a mistake by disallowing to exit from inputs of 'canonical' transaction when some other input is spent. I'm bringing that back in next commit.

File: plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py [hunk 3]
```diff
@@ -817,3 +817,117 @@ def start_ife_piggyback_and_process(spend_id, owner, testlang):
...
```
Comment (pgebal) [inline plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py]: I do not think that's harmful. Any other ideas?

File: plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py [hunk 4]
```diff
@@ -817,3 +817,117 @@ def start_ife_piggyback_and_process(spend_id, owner, testlang):
     testlang.piggyback_in_flight_exit_input(spend_id, 0, owner)
     testlang.forward_timestamp(2 * MIN_EXIT_PERIOD + 1)
     testlang.process_exits(NULL_ADDRESS, 0, 10)
+
+
+def test_should_be_resistant_to_Thomalla_heist(testlang, plasma_framework, token):
```
Comment (pdobacz) [inline plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py]: the heist name is jargon, let's be explicit here about what the scenario is

File: plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py (hunk 3 above)
Comment (pdobacz) [reply plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py]: if the piggybacks were valid (i.e. not attempting to double spend) per se, the bonds should be returned.

**Comment by boolafish** [comment]
Not 100% sure, but since new exit can override the finalization exitId, it seems like to be possible to have bug?

tx1: should be canonical, ETH and OMG
tx2: should be non-canonical due to input flagged by tx1

Now ETH of tx1 is process first. Since canonical, all inputs + ETH outputs would be flagged with `exitId` of tx1. Now tx2 ETH get processed. Since non-canonical (due to input flagged), it would re-flag the inputs with `exitId` of tx2. Not when OMG of tx1 get processed, it would be considered non-canonical.....

**Comment by pdobacz** [comment]
> but since new exit can override the finalization exitId

@boolafish not sure what you mean "can override"? But indeed this sounds dangerous. Is there an alicebobbing scenario that could illustrate the situation?

**Comment by boolafish** [comment]
The implementation looks like the later exit can flag the output with new exitId

**Comment by pgebal** [comment]
> Not 100% sure, but since new exit can override the finalization exitId, it seems like to be possible to have bug?
> 
> tx1: should be canonical, ETH and OMG
> tx2: should be non-canonical due to input flagged by tx1
> 
> Now ETH of tx1 is process first. Since canonical, all inputs + ETH outputs would be flagged with `exitId` of tx1. Now tx2 ETH get processed. Since non-canonical (due to input flagged), it would re-flag the inputs with `exitId` of tx2. Not when OMG of tx1 get processed, it would be considered non-canonical.....

right, exitId needs to be sticky, set only if prev. it was 0.

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 5]
```diff
@@ -113,30 +112,32 @@ library PaymentProcessInFlightExit {
         }
     }
 
-    function isAnyInputSpent(
+    // input is unspent for a given exitId if it's not marked in finalized outputIds map
+    // or it's marked with the same exitId
```
Comment (paulperegud) [inline plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: First line is rather obvious - follows from the name of the function. Second line's message is very not obvious - question pops up immediately: "why would it ever be marked by the same exitId?"

This is "what" comment, not "why" comment.

File: plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py (hunk 3 above)
Comment (paulperegud) [reply plasma_framework/python_tests/tests/contracts/root_chain/test_process_exits.py]: IMHO priorities should be in this order:
1. honest players never lose their bonds
2. prevent griefing
3. care about simplicity
4. punish wrongdoers

File: plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol [hunk 6]
```diff
@@ -67,10 +67,9 @@ library PaymentProcessInFlightExit {
 
         // Check whether any input is already spent. Required to prevent operator stealing funds.
         // See: https://github.com/omisego/plasma-contracts/issues/102#issuecomment-495809967
-        // Also, slightly different from the solution above, we treat input spent as non-canonical.
-        // So an IFE is only canonical if all inputs of the in-flight tx are not double spent by competing tx or exit.
-        // see: https://github.com/omisego/plasma-contracts/issues/470
-        if (!exit.isCanonical || isAnyInputSpent(self.framework, exit, token)) {
+        // Also, slightly different from the solution above
+        // see: https://github.com/omisego/security-issues/issues/13
```
[the remark this answers is not in the document]
Comment (paulperegud) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: @pgebal please document this solution as the comment in the code. Answering the question: "why we are doing this?"

[the remark this answers is not in the document]
Comment (pgebal) [reply plasma_framework/contracts/src/exits/payment/controllers/PaymentProcessInFlightExit.sol]: @paulperegud  I put some comments explaining why it's done. Please take a look at them. Are they clear?
@boolafish I removed that second link, I hope my comment is enough. If it's not, I can either expand the comment or make the issue public after it's fixed.

**Pull request merged**